ONSdigital/rm-collection-exercise-service
# 587 caching performance improvements

# Motivation and Context
To help improve the performance of the collection exercise validation and distribution, I've added cacheable annotations to the collection instrument and action plans endpoints to reduce the amount of calls it makes. From testing locally with a 5000 sample i've seen about a 1-2 minute improvement to validating the collection exercise and a slight improvement of the distribution to the case service. 

The performance could be even more improved by decreasing the `validation-schedule-delay-milli-seconds` and `distribution-schedule-delay-milli-seconds` time from 1 minute to something lower. Testing the effects this would have on the system locally is difficult, so it'd be best to test this in the performance environment when it's ready to go.

# What has changed 
- Added Cacheable annotations to `actionplans` and `collectioninstrument`

# How to test?
- In docker the `validation-schedule-delay-milli-seconds` and `distribution-schedule-delay-milli-seconds` are set to 1 second so to make it like production, you should change this to `60000`
- Run the tests
- Add some log lines inside `getActionPlanBySelectorsBusiness` and `requestCollectionInstruments` and execute a collection exercise, it should use a cached version and they evict the cache after about a minute. 

# Links
[Trello card](https://trello.com/c/lA9ni5qG)
